lib/Dialect/TritonGPU/Transforms/OptimizeDotOperands.cpp

```diff
@@ -192,6 +193,119 @@ class MoveOpAfterLayoutConversion : public mlir::RewritePattern {
 
 } // namespace
 
+static bool isConvertToDotEncoding(Operation *op) {
+  auto convertLayout = llvm::dyn_cast<ConvertLayoutOp>(op);
+  if (!convertLayout)
+    return false;
+  auto tensorType =
+      convertLayout.getResult().getType().cast<RankedTensorType>();
+  return tensorType.getEncoding().isa<DotOperandEncodingAttr>();
+}
+
+static ConvertLayoutOp updateConvert(OpBuilder &builder, ConvertLayoutOp cvt,
+                                     IRMapping &mapping, Type smallestType) {
+  auto cvtDstTy = cvt.getResult().getType().cast<RankedTensorType>();
+  auto cvtDstEnc = cvtDstTy.getEncoding().cast<DotOperandEncodingAttr>();
+  Value operand = cvt.getOperand();
+  if (mapping.contains(operand))
+    operand = mapping.lookup(operand);
+  auto newDstTy = RankedTensorType::get(
+      cvtDstTy.getShape(), cvtDstTy.getElementType(),
+      DotOperandEncodingAttr::get(cvtDstEnc.getContext(), cvtDstEnc.getOpIdx(),
+                                  cvtDstEnc.getParent(), smallestType));
+  auto newCvt =
+      builder.create<ConvertLayoutOp>(cvt.getLoc(), newDstTy, operand);
+  mapping.map(cvt.getResult(), newCvt.getResult());
+  return newCvt;
+}
+
+// Update kWidth based on the smallestType found in the given convert ops and
+// propagate the type change.
+static void
+updateDotEncodingLayout(SmallVector<ConvertLayoutOp> &convertsToDotEncoding,
+                        Type smallestType) {
+  IRMapping mapping;
+  OpBuilder builder(smallestType.getContext());
+  SetVector<Operation *> slices(convertsToDotEncoding.begin(),
+                                convertsToDotEncoding.end());
+  // Collect all the operations where the type needs to be propagated.
+  for (auto cvt : convertsToDotEncoding) {
+    auto filter = [&](Operation *op) {
+      for (Value operand : op->getOperands()) {
+        auto tensorType = operand.getType().dyn_cast<RankedTensorType>();
+        if (tensorType &&
+            tensorType.getEncoding().isa<DotOperandEncodingAttr>())
+          return true;
+      }
+      return false;
+    };
+    mlir::getForwardSlice(cvt.getResult(), &slices, {filter});
+  }
+  // Apply the type change by walking ops in toological order.
+  slices = mlir::topologicalSort(slices);
+  for (Operation *op : slices) {
+    builder.setInsertionPoint(op);
+    if (isConvertToDotEncoding(op)) {
+      auto cvt = cast<ConvertLayoutOp>(op);
+      ConvertLayoutOp newCvt =
+          updateConvert(builder, cvt, mapping, smallestType);
+      continue;
+    }
+    auto *newOp = cloneWithInferType(builder, op, mapping);
+    for (auto [result, newResult] :
+         llvm::zip(op->getResults(), newOp->getResults())) {
+      result.replaceUsesWithIf(newResult, [&](OpOperand &operand) {
+        return slices.count(operand.getOwner()) == 0;
+      });
+    }
+  }
+  for (Operation *op : llvm::reverse(slices))
```

Comment:
This is to erase users ops before a given op. Removing an op while it still has uses would assert.

Reply:
Seems to me that they can be removed in the CSE pass? Just to confirm

Reply:
you mean DCE? Yes. Do you think I should leave it to DCE instead?

Reply:
Yep. Sorry for the typo. 
Sure, I think leaving it to DCE should be ok

Reply:
Do we run DCE explicitly though? I don't see it being ran after that and before pipeliner?

Reply:
Simple DCE should be included in `applyPatternsAndFoldGreedily` invocations. 
It's OK if you just want to leave the erase here, in theory it should trigger problems.

Reply:
Yes this transformation doesn't use rewrite pattern infrastructure because it needs a more global decision (picking the minum type) so we don't get the DCE for free. I'll leave it for now then.